### ISSUE TITLE
(🎁) Intersections with `@final` types should trigger unreachable errors

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#15148]
```py
from typing import final

@final
class A: ...
class B: ...

a: A
assert isinstance(a, B)
reveal_type(a)  # note: Revealed type is "__main__.<subclass of "A" and "B">"
print("hi") # no unreachable error
```

**Comment by tyralla** [comment]
@KotlinIsland: #16330 should fix this issue.  Would you like to review it?
